Fix race condition in FontFile data access during threaded imports

d6e opened the pull request:
## Problems

1. **FontFile Data Race**

   * When importing fonts on multiple threads, two threads could grab the same `FontFile` from `ResourceCache`.
   * One thread calls `set_data()`, the other calls `get_data()` at the same time.
   * Copy-on-write kicks in and invalidates the pointer the second thread was using, leading to a crash.

2. **Resource Change Callbacks on Background Threads**

   * Import threads were firing resource-change callbacks directly.
   * That caused node operations off the main thread, triggering errors like:

     ```
     ERROR: Caller thread can't call this function in this node (/root). Use call_deferred() or call_thread_group() instead.
       at: propagate_notification (scene/main/node.cpp:2523)
     ```

---

## Fixes

### 1. FontFile Thread Safety

* **Files updated:** `scene/resources/font.h` and `scene/resources/font.cpp`
* **What changed:**

  * Added a `mutable Mutex data_mutex` inside `FontFile`.
  * Wrapped `set_data_ptr()`, `set_data()`, and `get_data()` in `data_mutex` locks.
  * In `set_data()`, forced a unique COW copy (`data.write()[0] = data[0]`) so no other thread is still holding an old pointer.
  * Added null checks in `get_data()` so it won’t try to read invalid data.

### 2. Resource Callback Thread Safety

* **File updated:** `core/io/resource_loader.cpp`
* **What changed:**

  * Swapped out `rcc.callable.call()` for `rcc.callable.call_deferred()` in `resource_changed_emit()`.
  * Now callbacks run on the main thread, avoiding those “can’t call this in a background thread” errors.

Commit:
Fix thread safety issue in resource changed callbacks during imports

Use call_deferred() instead of direct call() in ResourceLoader::resource_changed_emit() to ensure resource change callbacks execute on the main thread. This prevents segfaults when font imports trigger node operations from background threads.

clayjohn commented:
Do you happen to have an MRP so that PR reviewers can easily test your code and confirm that it works?

clayjohn commented:
Also should we close https://github.com/godotengine/godot/pull/107218 now since this PR fully supersedes it?

d6e commented:
> Do you happen to have an MRP so that PR reviewers can easily test your code and confirm that it works?

I don't have one, it's an issue that occurs in our very large project and only happens 1 out of approximately 20 times. Do I need one to continue with this?

d6e commented:
Okay, I think this MRP should do it. I can confirm it crashes without the fix and does not crash with the fix

[FontRaceConditionMRP.zip](https://github.com/user-attachments/files/20635592/FontRaceConditionMRP.zip)

Calinou reviewed (commented):
Tested locally on Windows 11 24H2 with the [MRP linked above](https://github.com/godotengine/godot/pull/107224#issuecomment-2951219498), it works as expected.

Note that I had to increase the `memory/limits/message_queue/max_size_mb` project setting in the MRP to get it to run correctly (I set it to `2048` just to be sure).

```
=== FontFile Race Condition Reproduction ===
This MRP uses real threading to reproduce the FontFile race condition
that causes crashes during concurrent font data access.

Created FontFile with test data, size: 2048
Starting threaded race condition test...
(This should crash in versions without the FontFile fix)

Stopping threads...

=== Results ===
UNEXPECTED: No crashes detected after 2275500 iterations
This suggests the FontFile race condition fix is already present,
or the race condition wasn't triggered this time.

Expected behavior:
- WITHOUT fix: Segfaults/crashes during execution
- WITH fix: Completes without crashes
```

However, in `master`, running the MRP didn't crash either, both with unoptimized and optimized editor builds compiled using MSVC 2022. I'm on a 9950X3D, so maybe I would need to adjust `OS.delay_msec()` calls to get the race condition to occur (assuming it was tested on a slower CPU).

d6e commented:
> @d6e Which OS and compiler are you using to reproduce the issue (and test the fix)?

Linux and gcc 13.3.0

Commit:
Fix race condition in FontFile data access during threaded imports

Add mutex protection to FontFile data access methods to prevent
race conditions when multiple import threads access the same
cached FontFile simultaneously.

Use data_external flag to track whether data came from set_data_ptr()
vs set_data() to avoid unnecessary PackedByteArray COW copies.

AThousandShips commented on scene/resources/font.h:
```diff
@@ -31,6 +31,7 @@
 #pragma once
 
 #include "core/io/resource.h"
+#include "core/os/mutex.h"
```
```suggestion
```
Already included

d6e committed:
Remove include from scene/resources/font.h

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>
